src/bluetooth_numbers/reverse_lookup.py

```diff
@@ -0,0 +1,122 @@
+UUID_TYPE_DEFAULT = ("characteristic", "company", "descriptor", "oui", "service")
+
+
+class Match(NamedTuple):
+    """Named tuple to hold a UUID and its description."""
+
+    uuid: str | UUID | int
+    description: str
+    uuid_type: str
+
+
+class ReverseLookup:
+    """Reverse lookup class to find UUIDs by their description.
+
+    Examples:
+            >>> from bluetooth_numbers.reverse_lookup import ReverseLookup, Match
+            >>> rl = ReverseLookup()
+            >>> matches = rl.lookup("Cycling Power")
+            >>> Match('00:05:5A', 'Power Dsine Ltd.', 'oui') in matches
+            True
+            >>> rl.lookup("Cycling Power", logic="AND")
+            {Match(uuid=6168, description='Cycling Power', uuid_type='service'),
+             Match(uuid=10851, description='Cycling Power Measurement',
+             uuid_type='characteristic'),
+             Match(uuid=10852, description='Cycling Power Vector',
+             uuid_type='characteristic'),
+             Match(uuid=10853, description='Cycling Power Feature',
+             uuid_type='characteristic'),
+             Match(uuid=10854, description='Cycling Power Control Point',
+             uuid_type='characteristic')}
+            >>> rl.lookup("Power Feature", uuid_types=['characteristic'],
+            logic="SUBSTR")
+            {Match(uuid=10853, description='Cycling Power Feature',
+            uuid_type='characteristic')}
+    """
+
+    def __init__(self) -> None:
+        """Initialize the ReverseLookup class, build index."""
+        self.index = self._build_index()
+
+    def _build_index(self) -> dict:
+        """Build dictionary (index) of terms to UUIDs.
+
+        Returns:
+            dict: Dictionary of terms to UUIDs.
+        """
+        reverse_lookup = {}
+        uuid_dicts = (
+            (characteristic, "characteristic"),
+            (company, "company"),
+            (descriptor, "descriptor"),
+            (oui, "oui"),
+            (service, "service"),
+        )
+        for uuid_dict, uuid_type in uuid_dicts:
+            for uuid, description in uuid_dict.items():
+                for term in description.lower().split(" "):
+                    if term not in reverse_lookup:
+                        reverse_lookup[term] = set()
+                    reverse_lookup[term].add(Match(uuid, description, uuid_type))
+        return reverse_lookup
+
+    def lookup(
+        self,
+        terms: str,
+        uuid_types: list[UUID_TYPES] = UUID_TYPE_DEFAULT,
+        logic: LOGIC = "OR",
+    ) -> set:
+        """Return the UUIDs for a given term(s).
+
+        Args:
+            terms: String with the term(s) to search for.
+            uuid_types: List of UUID types to search in.
+            logic: Search logic to use. Can be "OR", "AND" or "SUBSTR".
+
+        Returns:
+            set or Match, named tuples, (uuid, description, uuid_type)
+        """
+        terms_set = set(terms.lower().split(" "))
+        results = set()
```

Comment:
Let's add `# type: ignore[var-annotated]`, we don't need to make mypy 100% happy.